[WPE][GTK] Instrument WebMemorySampler on Linux

GeorgesStavracas opened the pull request:
#### 0322ec1d349a684c586b25c53c15de9478ce975e
<pre>
[WPE][GTK] Instrument WebMemorySampler on Linux
<a href="https://bugs.webkit.org/show_bug.cgi?id=298205">https://bugs.webkit.org/show_bug.cgi?id=298205</a>

Reviewed by Michael Catanzaro.

Use WTFSetCounter() in conjunction to WebMemorySamplerLinux to proxy
the metrics to Sysprof. The WEBKIT_SAMPLE_MEMORY env var enables this.

* Source/WebKit/Shared/linux/WebMemorySamplerLinux.cpp:
(WebKit::WebMemorySampler::sampleWebKit const):

Canonical link: <a href="https://commits.webkit.org/299428@main">https://commits.webkit.org/299428@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7cdcdcacc44a5170b0750f6fcbe46cff03bd2da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118936 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29268 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125137 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70999 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39313 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47199 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90260 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121889 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106633 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70765 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30399 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68789 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100783 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24935 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128176 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45843 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34631 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98922 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102853 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98701 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25099 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44159 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22159 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42408 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45713 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51391 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45179 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48523 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46863 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e91b5e835ee95d86158b55eefe6017bf01452aa5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50157

GeorgesStavracas commented:
How it looks:

<img width="2640" height="3862" alt="image" src="https://github.com/user-attachments/assets/1583e3d7-38a9-4d2f-8e3a-615b754c5fc7" />

GeorgesStavracas commented on Source/WebKit/Shared/linux/WebMemorySamplerLinux.cpp:
```diff
@@ -77,6 +78,12 @@ static inline void appendKeyValuePair(WebMemoryStatistics& stats, const String&
     stats.values.append(value);
 }
 
+#define INSTRUMENT_KEY_VALUE_COUNTER(stats, id, key, value) \
+do { \
+    WTFSetCounter(id, value); \
+    appendKeyValuePair(stats, key, value); \
+} while (0);
```
I need the macro because `WTFSetCounter` stringifies the counter id

GeorgesStavracas committed:
[WPE][GTK] Instrument WebMemorySampler on Linux
https://bugs.webkit.org/show_bug.cgi?id=298205

Reviewed by Michael Catanzaro.

Use WTFSetCounter() in conjunction to WebMemorySamplerLinux to proxy
the metrics to Sysprof. The WEBKIT_SAMPLE_MEMORY env var enables this.

* Source/WebKit/Shared/linux/WebMemorySamplerLinux.cpp:
(WebKit::WebMemorySampler::sampleWebKit const):

Canonical link: https://commits.webkit.org/299428@main
